appinventor/buildserver/src/com/google/appinventor/buildserver/FormPropertiesAnalyzer.java

```diff
@@ -160,87 +158,62 @@ private static void collectComponentTypes(JSONObject componentProperties,
    * @return A mapping of component type names to sets of blocks used
    */
   public static Map<String, Set<String>> getComponentBlocksFromBlocksFile(String source) {
-    Map<String, Set<String>> result = new HashMap<>();
+    final Map<String, Set<String>> result = new HashMap<>();
     if (source == null || source.isEmpty()) {
       return result;
     }
     try {
-      DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
-      DocumentBuilder builder = factory.newDocumentBuilder();
-      Document doc = builder.parse(new ByteArrayInputStream(source.getBytes(StandardCharsets.UTF_8)));
-      NodeList blocks = doc.getElementsByTagName("block");
-      for (int i = 0; i < blocks.getLength(); i++) {
-        Element block = (Element) blocks.item(i);
-        String type = block.getAttribute("type");
-
-        // Skip non-component and disabled blocks
-        if (!isComponentBlock(type) || "true".equals(block.getAttribute("disabled"))) continue;
-
-        Element mutation = getMutation(block);
-        if (mutation != null) {  // Should always be true, in theory...
-          String componentType = mutation.getAttribute("component_type");
-          String blockName = getBlockName(type, mutation);
-          if (!result.containsKey(componentType)) {
-            result.put(componentType, new HashSet<String>());
+      XMLReader reader = XMLReaderFactory.createXMLReader();
+      reader.setContentHandler(new DefaultHandler() {
+        private int skipBlocksCounter = 0;
+        private String blockType;
+
+        @Override
+        public void startElement(String uri, String localName, String qName, Attributes attributes)
+            throws SAXException {
+          if ("block".equals(qName)) {
```

Comment:
One last corner case (hopefully): Because `<next>` elements are contained with `<block>` elements, once you have a disabled block, any blocks following it are considered disabled even though they are logically separate statements. This results in the following block stack not generating the SEND_SMS permission even though SendMessageDirect is enabled.

<img width="336" alt="Screen Shot 2019-03-22 at 10 32 42 AM" src="https://user-images.githubusercontent.com/1331253/54830233-80b62b00-4c8e-11e9-8244-45df5c9c8431.png">

My first thought was to reset the `skipBlocksCounter` to 0 when we encounter the `<next>` tag, but that won't work because the `<next>` might be associated with a value or statement input rather than the disabled block. Instead, when we encounter a disabled `<block>` we should add an entry to a stack (could be a `LinkedList`) and update the `skipBlocksCounter` for that entry. We pop the element off the stack if `skipBlockCounter` is 0 and either we hit an opening `<next>` tag or a closing `</block>` tag.

Reply:
I have never thought `<next>` is in `<block>`,

Reply:
![image](https://user-images.githubusercontent.com/3827464/54861209-8aa85e80-4d60-11e9-8f3c-7f782b6701e0.png)

Test result:

1 Sharing:ShareMessage
2 PhoneCall:MakePhoneCall
3 NxtDrive:Stop
4 NxtDrive:TurnClockwiseIndefinitely
5 NxtColorSensor:GetColor